include/cuco/detail/hyperloglog/storage.cuh

```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cuda/std/array>
+
+namespace cuco::detail {
+template <int32_t Precision>
+struct alignas(sizeof(int) * 4) hyperloglog_storage
+  : public cuda::std::array<int, 1ull << Precision> {
```

Comment:
comment: Inheriting from a standard library type is usually a code smell. It's almost always better to use composition instead of inheritance.

Reply:
Good point. I'll go with composition then and expose `operator[]` and `size()` manually since these are the only APIs required for the storage.

Reply:
I will probably also move some of the functionality like merging histograms to the storage class itself so we get a better separation of responsibilities. The function for updating the max count of a register is also a candidate.

Reply:
See commit: [cea2afb](https://github.com/NVIDIA/cuCollections/pull/429/commits/cea2afb36f1942f925501041463adda95348ae0f)

Reply:
> Are we relying on any improvement brought by the manual alignment? Otherwise, the whole class is not needed but a type alias in the hyperloglog class.

Sorry, @PointKernel I missed that one. For the dense storage type we could go with a simple typedef. However, if we want to enable the sparse representation (basically a `static_reduction_map`) in order to improve the memory density even further, we would need this as a separate class. The sparse representation would also pair nicely with NVIDIA/cuCollections#384 since we need a `max` reduction over payloads. This would allow us to fit estimators with higher `Precision` values in shmem and run close to SOL. I haven't tested it yet, but it should be clear that once we need to spill the registers to local memory the performance is going to drop drastically.